rootkit_lrk5/findutils-4.4.0/locate/updatedb.1
.TH UPDATEDB 1 \" -*- nroff -*-
.SH NAME
updatedb \- update a file name database
.SH SYNOPSIS
.B updatedb [\fIoptions\fP]
.SH DESCRIPTION
This manual page
documents the GNU version of
.BR updatedb ,
which updates file name databases used by GNU
.BR locate .
The file name databases contain lists of files that were in
particular directory trees when the databases were last updated.
The file name of the default database is determined when \fBlocate\fP
and \fBupdatedb\fP are configured and installed.  The frequency with
which the databases are updated and the directories for which they
contain entries depend on how often \fBupdatedb\fP is run, and with
which arguments.
.P
In networked environments, it often makes sense to build a database at
the root of each filesystem, containing the entries for that filesystem.
.B updatedb
is then run for each filesystem on the fileserver where that
filesystem is on a local disk, to prevent thrashing the network.
Users can select which databases \fBlocate\fP searches using an
environment variable or command line option; see \fBlocate\fP(1).
Databases can not be concatenated together.
.P
The file name database format changed starting with GNU
.B find
and
.B locate
version 4.0 to allow machines with different byte orderings to share
the databases.  The new GNU
.B locate
can read both the old and new database formats.
However, old versions of
.B locate
and
.B find
produce incorrect results if given a new-format database.
.SH OPTIONS
.TP
.B \-\-findoptions='\fI\-option1 \-option2...\fP'
Global options to pass on to \fBfind\fP.
The environment variable \fBFINDOPTIONS\fP also sets this value.
Default is none.
.TP
.B \-\-localpaths='\fIpath1 path2...\fP'
Non-network directories to put in the database.
Default is /.
.TP
.B \-\-netpaths='\fIpath1 path2...\fP'
Network (NFS, AFS, RFS, etc.) directories to put in the database.
The environment variable \fBNETPATHS\fP also sets this value.
Default is none.
.TP
.B \-\-prunepaths='\fIpath1 path2...\fP'
Directories to not put in the database, which would otherwise be.
Remove any trailing slashes from the path names, otherwise 
.B updatedb
won\'t recognise the paths you want to omit (because it uses them as
regular expression patterns).
The environment variable \fBPRUNEPATHS\fP also sets this value.
Default is /tmp /usr/tmp /var/tmp /afs.
.TP
.B \-\-prunefs='\fIpath...\fP'
File systems to not put in the database, which would otherwise be.
Note that files are pruned when a file system is reached;
Any file system mounted under an undesired file system will be
ignored.
The environment variable
\fBPRUNEFS\fP also sets this value.
Default is \fInfs NFS proc\fP.
.TP
.B \-\-output=\fIdbfile\fP
The database file to build.
Default is system-dependent.  In Debian GNU/Linux, the default
is /var/cache/locate/locatedb.
.TP
.B \-\-localuser=\fIuser\fP
The user to search non-network directories as, using \fBsu\fP(1).
Default is to search the non-network directories as the current user.
You can also use the environment variable \fBLOCALUSER\fP to set this user.
.TP
.B \-\-netuser=\fIuser\fP
The user to search network directories as, using \fBsu\fP(1).
Default is \fBdaemon\fP.
You can also use the environment variable \fBNETUSER\fP to set this user.
.TP
.B \-\-old\-format
Create the database in the old format.  This is a synonym for 
.BR \-\-dbformat=old .
.TP
.B \-\-dbformat=F
Create the database in format F.  The default format is called LOCATE02.
F can be 
.B old
to select the old database format (this is the same as specifying
.BR \-\-old\-format ).
Alternatively the 
.B slocate
format is also supported.  When the 
.B slocate 
format is in use, the database produced is marked as having security
level 1.  If you want to build a system-wide
.B slocate
database, you may want to run 
.B updatedb
as root.
.TP
.B \-\-version
Print the version number of
.B updatedb
and exit.
.TP
.B "\-\-help"
Print a summary of the options to
.B updatedb
and exit.
.SH "SEE ALSO"
\fBfind\fP(1), \fBlocate\fP(1), \fBlocatedb\fP(5), \fBxargs\fP(1)
\fBFinding Files\fP (on-line in Info, or printed)
.SH "BUGS"
.P
The 
.B updatedb
program correctly handles filenames containing newlines,
but only if the system's sort command has a working
.I \-z
option.  If you suspect that 
.B locate
may need to return filenames containing newlines, consider using its
.I \-\-null 
option.
.P
The best way to report a bug is to use the form at
http://savannah.gnu.org/bugs/?group=findutils.  
The reason for this is that you will then be able to track progress in
fixing the problem.   Other comments about \fBupdatedb\fP(1) and about
the findutils package in general can be sent to the 
.I bug-findutils
mailing list.  To join the list, send email to 
.IR bug-findutils-request@gnu.org .


rootkit_lrk5/findutils-4.4.0/locate/updatedb.sh
#! /bin/sh
# updatedb -- build a locate pathname database
# Copyright (C) 1994, 1996, 1997, 2000, 2001, 2003, 2004, 2005, 2006
# Free Software Foundation, Inc.
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
# 
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
# 
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.

# csh original by James Woods; sh conversion by David MacKenzie.

#exec 2> /tmp/updatedb-trace.txt 
#set -x

version='
updatedb (@PACKAGE_NAME@) @VERSION@
Copyright (C) 2007 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.

Written by Eric B. Decker, James Youngman, and Kevin Dalley.
'


usage="\
Usage: $0 [--findoptions='-option1 -option2...']
       [--localpaths='dir1 dir2...'] [--netpaths='dir1 dir2...']
       [--prunepaths='dir1 dir2...'] [--prunefs='fs1 fs2...']
       [--output=dbfile] [--netuser=user] [--localuser=user] 
       [--old-format] [--dbformat] [--version] [--help]

Report bugs to <bug-findutils@gnu.org>."
changeto=/
old=no
for arg
do
  # If we are unable to fork, the back-tick operator will 
  # fail (and the shell will emit an error message).  When 
  # this happens, we exit with error value 71 (EX_OSERR).
  # Alternative candidate - 75, EX_TEMPFAIL.
  opt=`echo $arg|sed 's/^\([^=]*\).*/\1/'`  || exit 71
  val=`echo $arg|sed 's/^[^=]*=\(.*\)/\1/'` || exit 71
  case "$opt" in
    --findoptions) FINDOPTIONS="$val" ;;
    --localpaths) SEARCHPATHS="$val" ;;
    --netpaths) NETPATHS="$val" ;;
    --prunepaths) PRUNEPATHS="$val" ;;
    --prunefs) PRUNEFS="$val" ;;
    --output) LOCATE_DB="$val" ;;
    --netuser) NETUSER="$val" ;;
    --localuser) LOCALUSER="$val" ;;
    --old-format) old=yes ;;
    --changecwd)  changeto="$val" ;;
    --dbformat)   dbformat="$val" ;;
    --version) fail=0; echo "$version" || fail=1; exit $fail ;;
    --help)    fail=0; echo "$usage"   || fail=1; exit $fail ;;
    *) echo "updatedb: invalid option $opt
$usage" >&2
       exit 1 ;;
  esac
done




case "${dbformat:+yes}_${old}" in 
    yes_yes)
	echo "The --dbformat and --old cannot both be specified." >&2
	exit 1
	;;
	*) 
	;;
esac

if test "$old" = yes || test "$dbformat" = "old" ; then
    echo "Warning: future versions of findutils will shortly discontinue support for the old locate database format." >&2
    old=yes
    sort="@SORT@"
    print_option="-print"
    frcode_options=""
else
    frcode_options=""
    case "$dbformat" in 
	"")
		# Default, use LOCATE02
	    ;;
	LOCATE02)
	    ;;
	slocate)
	    frcode_options="$frcode_options -S 1"
	    ;;
	*)
	    echo "Unsupported locate database format ${dbformat}: Supported formats are:" >&2
	    echo "LOCATE02, slocate, old" >&2
	    exit 1
    esac


    if @SORT_SUPPORTS_Z@
    then
        sort="@SORT@ -z"
        print_option="-print0"
        frcode_options="$frcode_options -0"
    else
        sort="@SORT@"
        print_option="-print"
    fi
fi

getuid() {
    # format of "id" output is ...
    # uid=1(daemon) gid=1(other)
    # for `id's that don't understand -u
    id | cut -d'(' -f 1 | cut -d'=' -f2
}

# figure out if su supports the -s option
select_shell() {
    if su "$1" -s $SHELL -c false < /dev/null  ; then
	# No.
	echo ""
    else
	if su "$1" -s $SHELL -c true < /dev/null  ; then
	    # Yes.
	    echo "-s $SHELL"
        else
	    # su is unconditionally failing.  We won't be able to 
	    # figure out what is wrong, so be conservative.
	    echo ""
	fi
    fi
}


# You can set these in the environment, or use command-line options,
# to override their defaults:

# Any global options for find?
: ${FINDOPTIONS=}

# What shell shoud we use?  We should use a POSIX-ish sh.
: ${SHELL="/bin/sh"}

# Non-network directories to put in the database.
: ${SEARCHPATHS="/"}

# Network (NFS, AFS, RFS, etc.) directories to put in the database.
: ${NETPATHS=}

# Directories to not put in the database, which would otherwise be.
: ${PRUNEPATHS="/tmp /usr/tmp /var/tmp /afs /amd /sfs /proc"}

# Trailing slashes result in regex items that are never matched, which 
# is not what the user will expect.   Therefore we now reject such 
# constructs.
for p in $PRUNEPATHS; do
    case "$p" in
	/*/)   echo "$0: $p: pruned paths should not contain trailing slashes" >&2
	       exit 1
    esac
done

# The same, in the form of a regex that find can use.
test -z "$PRUNEREGEX" &&
  PRUNEREGEX=`echo $PRUNEPATHS|sed -e 's,^,\\\(^,' -e 's, ,$\\\)\\\|\\\(^,g' -e 's,$,$\\\),'`

# The database file to build.
: ${LOCATE_DB=@LOCATE_DB@}

# Directory to hold intermediate files.
if test -d /var/tmp; then
  : ${TMPDIR=/var/tmp}
elif test -d /usr/tmp; then
  : ${TMPDIR=/usr/tmp}
else
  : ${TMPDIR=/tmp}
fi
export TMPDIR

# The user to search network directories as.
: ${NETUSER=daemon}

# The directory containing the subprograms.
if test -n "$LIBEXECDIR" ; then
    : LIBEXECDIR already set, do nothing
else
    : ${LIBEXECDIR=@libexecdir@}
fi

# The directory containing find.
if test -n "$BINDIR" ; then
    : BINDIR already set, do nothing
else
    : ${BINDIR=@bindir@}
fi

# The names of the utilities to run to build the database.
: ${find:=${BINDIR}/@find@}
: ${frcode:=${LIBEXECDIR}/@frcode@}
: ${bigram:=${LIBEXECDIR}/@bigram@}
: ${code:=${LIBEXECDIR}/@code@}


checkbinary () {
    if test -x "$1" ; then
	: ok
    else
      eval echo "updatedb needs to be able to execute $1, but cannot." >&2
      exit 1
    fi
}

for binary in $find $frcode $bigram $code
do
  checkbinary $binary
done


PATH=/bin:/usr/bin:${BINDIR}; export PATH

: ${PRUNEFS="nfs NFS proc afs proc smbfs autofs iso9660 ncpfs coda devpts ftpfs devfs mfs sysfs shfs"}

if test -n "$PRUNEFS"; then
prunefs_exp=`echo $PRUNEFS |sed -e 's/\([^ ][^ ]*\)/-o -fstype \1/g' \
 -e 's/-o //' -e 's/$/ -o/'`
else
  prunefs_exp=''
fi

# Make and code the file list.
# Sort case insensitively for users' convenience.

rm -f $LOCATE_DB.n
trap 'rm -f $LOCATE_DB.n; exit' HUP TERM

if test $old = no; then
# LOCATE02 or slocate format
if {
cd "$changeto"
if test -n "$SEARCHPATHS"; then
  if [ "$LOCALUSER" != "" ]; then
    # : A1
    su $LOCALUSER `select_shell $LOCALUSER` -c \
    "$find $SEARCHPATHS $FINDOPTIONS \
     \\( $prunefs_exp \
     -type d -regex '$PRUNEREGEX' \\) -prune -o $print_option"
  else
    # : A2
    $find $SEARCHPATHS $FINDOPTIONS \
     \( $prunefs_exp \
     -type d -regex "$PRUNEREGEX" \) -prune -o $print_option
  fi
fi

if test -n "$NETPATHS"; then
myuid=`getuid` 
if [ "$myuid" = 0 ]; then
    # : A3
    su $NETUSER `select_shell $NETUSER` -c \
     "$find $NETPATHS $FINDOPTIONS \\( -type d -regex '$PRUNEREGEX' -prune \\) -o $print_option" ||
    exit $?
  else
    # : A4
    $find $NETPATHS $FINDOPTIONS \( -type d -regex "$PRUNEREGEX" -prune \) -o $print_option ||
    exit $?
  fi
fi
} | $sort -f | $frcode $frcode_options > $LOCATE_DB.n
then
    : OK so far
    true
else
    rv=$?
    echo "Failed to generate $LOCATE_DB.n" >&2
    rm -f $LOCATE_DB.n
    exit $rv
fi

# To avoid breaking locate while this script is running, put the
# results in a temp file, then rename it atomically.
if test -s $LOCATE_DB.n; then
  chmod 644 ${LOCATE_DB}.n
  mv ${LOCATE_DB}.n $LOCATE_DB
else
  echo "updatedb: new database would be empty" >&2
  rm -f $LOCATE_DB.n
fi

else # old

if ! bigrams=`mktemp -t updatedbXXXXXXXXX`; then
    echo mktemp failed >&2
    exit 1
fi

if ! filelist=`mktemp -t updatedbXXXXXXXXX`; then
    echo mktemp failed >&2
    exit 1
fi

rm -f $LOCATE_DB.n
trap 'rm -f $bigrams $filelist $LOCATE_DB.n; exit' HUP TERM

# Alphabetize subdirectories before file entries using tr.  James Woods says:
# "to get everything in monotonic collating sequence, to avoid some
# breakage i'll have to think about."
{
cd "$changeto"
if test -n "$SEARCHPATHS"; then
  if [ "$LOCALUSER" != "" ]; then
    # : A5
    su $LOCALUSER `select_shell $LOCALUSER` -c \
    "$find $SEARCHPATHS $FINDOPTIONS \
     \( $prunefs_exp \
     -type d -regex '$PRUNEREGEX' \) -prune -o $print_option" || exit $?
  else
    # : A6
    $find $SEARCHPATHS $FINDOPTIONS \
     \( $prunefs_exp \
     -type d -regex "$PRUNEREGEX" \) -prune -o $print_option || exit $?
  fi
fi

if test -n "$NETPATHS"; then
  myuid=`getuid`
  if [ "$myuid" = 0 ]; then
    # : A7
    su $NETUSER `select_shell $NETUSER` -c \
     "$find $NETPATHS $FINDOPTIONS \\( -type d -regex '$PRUNEREGEX' -prune \\) -o $print_option" ||
    exit $?
  else
    # : A8
    $find $NETPATHS $FINDOPTIONS \( -type d -regex "$PRUNEREGEX" -prune \) -o $print_option ||
    exit $?
  fi
fi
} | tr / '\001' | $sort -f | tr '\001' / > $filelist

# Compute the (at most 128) most common bigrams in the file list.
$bigram $bigram_opts < $filelist | sort | uniq -c | sort -nr |
  awk '{ if (NR <= 128) print $2 }' | tr -d '\012' > $bigrams

# Code the file list.
$code $bigrams < $filelist > $LOCATE_DB.n

rm -f $bigrams $filelist

# To reduce the chances of breaking locate while this script is running,
# put the results in a temp file, then rename it atomically.
if test -s $LOCATE_DB.n; then
  chmod 644 ${LOCATE_DB}.n
  mv ${LOCATE_DB}.n $LOCATE_DB
else
  echo "updatedb: new database would be empty" >&2
  rm -f $LOCATE_DB.n
fi

fi

exit 0
